src/app/codegen-interaction-model/CodegenDataModel.cpp

```diff
@@ -0,0 +1,411 @@
+    }
+
+    const EmberAfCluster * cluster = emberAfFindServerCluster(path.mEndpointId, path.mClusterId);
+    if (cluster != nullptr)
+    {
+        mPreviouslyFoundCluster = std::make_optional<ClusterReference>(path, cluster);
+    }
+    return cluster;
+}
+
+InteractionModel::AttributeEntry CodegenDataModel::NextAttribute(const ConcreteAttributePath & before)
+{
+    const EmberAfCluster * cluster = FindServerCluster(before);
+    VerifyOrReturnValue(cluster != nullptr, InteractionModel::AttributeEntry::Invalid());
+    VerifyOrReturnValue(cluster->attributeCount > 0, InteractionModel::AttributeEntry::Invalid());
+    VerifyOrReturnValue(cluster->attributes != nullptr, InteractionModel::AttributeEntry::Invalid());
+
+    // Handles global attribute iteration: if we got a global attribute, move to the next
+    switch (before.mAttributeId)
+    {
+    case Clusters::Globals::Attributes::GeneratedCommandList::Id:
+        return AttributeEntryForGlobalListAttribute(before, Clusters::Globals::Attributes::AcceptedCommandList::Id);
+    case Clusters::Globals::Attributes::AcceptedCommandList::Id:
+#if CHIP_CONFIG_ENABLE_EVENTLIST_ATTRIBUTE
+        return AttributeEntryForGlobalListAttribute(before, Clusters::Globals::Attributes::EventList::Id);
+    case Clusters::Globals::Attributes::EventList::Id:
+#endif // CHIP_CONFIG_ENABLE_EVENTLIST_ATTRIBUTE
+        return AttributeEntryForGlobalListAttribute(before, Clusters::Globals::Attributes::AttributeList::Id);
+    case Clusters::Globals::Attributes::AttributeList::Id:
+        return InteractionModel::AttributeEntry::Invalid();
+    default:
+        // pass-through: not a global attribute, try to find a "regular" attribute
+        break;
+    }
+
+    // find the given attribute in the list and then return the next one
+    std::optional<unsigned> attribute_idx = TryFindAttributeIndex(cluster, before.mAttributeId);
+    if (!attribute_idx.has_value())
+    {
+        return InteractionModel::AttributeEntry::Invalid();
+    }
+
+    unsigned next_idx = *attribute_idx + 1;
+    if (next_idx < cluster->attributeCount)
+    {
+        mAttributeIterationHint = next_idx;
+        return AttributeEntryFrom(before, cluster->attributes[next_idx]);
+    }
+
+    // We reach here if next_idx is just past the last attribute metadata. Return the first global
+    // attribute
+    return AttributeEntryForGlobalListAttribute(before, Clusters::Globals::Attributes::GeneratedCommandList::Id);
+}
+
+std::optional<InteractionModel::AttributeInfo> CodegenDataModel::GetAttributeInfo(const ConcreteAttributePath & path)
+{
+    const EmberAfCluster * cluster = FindServerCluster(path);
+
+    VerifyOrReturnValue(cluster != nullptr, std::nullopt);
+    VerifyOrReturnValue(cluster->attributeCount > 0, std::nullopt);
+    VerifyOrReturnValue(cluster->attributes != nullptr, std::nullopt);
+
+    switch (path.mAttributeId)
+    {
+    case Clusters::Globals::Attributes::GeneratedCommandList::Id:
```

Comment:
Again, we should handle the global list attributes outside here, so data model impls don't all have to have this chunk of code.

Reply:
See above: I think this separation of concerns is better. Letting upper layer handle global attributes causes a clearer separation of concerns.

I also imagine we will have some base class or helper that will indeed list global attributes as it does make sense for them to exist. However I believe they would be coupled with also providing metadata for them, so that said base/helper class can implement both iterate and read.